std/range/primitives.d

```diff
@@ -272,6 +272,20 @@ private void doPut(R, E)(ref R r, auto ref E e)
             "Cannot put a " ~ E.stringof ~ " into a " ~ R.stringof ~ ".");
         r.put(e);
     }
+    else static if (isNarrowString!R && is(const(E) == const(typeof(r[0]))))
+    {
+        // one character, we can put it
+        r[0] = e;
+        r = r[1 .. $];
+    }
+    else static if (isNarrowString!R && isNarrowString!E && is(typeof(r[] = e)))
+    {
+        // slice assign. Note that this is a duplicate from put, but because
+        // putChar uses doPut exclusively, we have to copy it here.
+        immutable len = e.length;
+        r[0 .. len] = e;
+        r = r[len .. $];
+    }
```

Comment:
This made it work, but I'm open to moving stuff around. I'm not quite sure what the point of doPut is, but I think this follows the rules above (it doesn't slice or modify `e` at all).

Reply:
On line 379, if you change the static if, you should be able to consolidate this code with the existing code.

Reply:
So that is what I did originally (I just removed the !isNarrowString). But the problem is that if the char widths don't match, it defers to `putChar`, but `putChar` never looks back at `put` again, it only goes to `doPut`. So I had to copy that code here.

Or maybe you mean something different?

I'm not sure the reason for this whole `put` trifecta, it's really hard to tell the purpose for all of these. So I figured this was the least disruptive.

Reply:
Just add the condition on line 281 as an OR to the condition on line 379

Reply:
Oh, I see the issue. Yeah you'll have to add an `is(typeof(r[0]) == typeof(e[0]))`

Reply:
But then it won't work for putting one type of character into another type of buffer. For example, putting a `wchar` into a `char[]`. Right now, that does work.